src/local.py

```diff
@@ -1,84 +1,113 @@
+import os, logging
 
-import sys
-if sys.platform == 'win32':
+from consts import (
+    GameID,
+    REGISTRY_START_PATHS,
+    SOFTWARE_PATHS,
+    GAME_REGISTY_RELATIVE_LOCATIONS,
+    REGISTRY_EXE_KEYS,
+    WIN_UNINSTALL_RELATIVE_LOCATION,
+    IS_WINDOWS,
+)
+import utils
+
+
+if IS_WINDOWS:
     import winreg
 
-from consts import MINECRAFT_REGISTRY_PATHS, REGISTRY_START_PATHS, MINECRAFT_REGISTRY_PATH_INSTALL_LOCATION_KEY, WINDOWS_UNINSTALL_LOCATION
+log = logging.getLogger(__name__)
 
-import logging as log
-import os
-import psutil
-import asyncio
 
-class LocalClient():
+class LocalClient:
     def __init__(self):
-        self.running_process = None
-
-    async def get_size_at_path(self, start_path):
-        total_size = 0
-        for dirpath, dirnames, filenames in os.walk(start_path):
-            for f in filenames:
-                await asyncio.sleep(0)
-                fp = os.path.join(dirpath, f)
-                if not os.path.islink(fp):
-                    total_size += os.path.getsize(fp)
-
-        return total_size
-
-    def get_minecraft_launcher_path(self):
-        if sys.platform == 'win32':
-            install_path = self.get_minecraft_folder_path()
-            potential_path = os.path.join(install_path, 'MinecraftLauncher.exe')
-            if os.path.exists(potential_path):
-                return potential_path
-        else:
-            return self.get_minecraft_folder_path()
-
-    def get_minecraft_folder_path(self):
-        if sys.platform == 'win32':
-            for start_path in REGISTRY_START_PATHS:
-                for minecraft_reg_path in MINECRAFT_REGISTRY_PATHS:
-                    try:
-                        reg = winreg.ConnectRegistry(None, start_path)
-                        with winreg.OpenKey(reg, minecraft_reg_path) as key:
-                            install_path = winreg.QueryValueEx(key, MINECRAFT_REGISTRY_PATH_INSTALL_LOCATION_KEY)[0]
-                    except OSError:
-                        continue
-                    if os.path.exists(install_path):
-                        return install_path
-        else:
-            potential_path = "/Applications/Minecraft.app"
-            if os.path.exists(potential_path):
-                return potential_path
+        self.running_games = {
+            GameID.Minecraft: None,
+            GameID.MinecraftDungeons: None,
+        }
+
+    def find_launcher_path(self, game_id, *, folder=False, folder_path=None, exe=None) -> str:
+        if exe is None and folder_path is not None:
+            folder_path = None
+        if exe is not None and folder_path is None:
+            folder_path = exe

```

Comment:
Yeah sorry, the code for that is a little messy over the multiple classes. Basically, that line is for Mac when the subclass returns the path to `Minecraft.app`. The thing is that's both the executable and the folder! That's why it's necessary but I could probably clean it up a little.

Reply:
Ou, I see. In this case the best would be to inject OS implementation to LocalClient (composition) and expose only higher level methods in LocalClient (like `run` or whatever you need there). Or, to avoid refactor just put some "if" statements. It is better to have more code but explicite.

```python
if WINDOWS:
    do_this()
else:
    do_that()
```

Reply:
every folder is executable BTW. "Running" .app folders on Mac is just a syntastic sugar for system layer to run proper executable `x.app/Contents/MacOS`

Reply:
> In this case the best would be to inject OS implementation to LocalClient (composition) and expose only higher level methods in LocalClient (like `run` or whatever you need there).

Yes, I definitely need to refactor local client to be better, in fact I kind of started halfway with the `launch` functions but then stopped because it was working. I'll work on it anyway :)

> "Running" .app folders on Mac is just a syntastic sugar for system layer to run proper executable in `x.app/Contents/MacOS`.

Absolutely did not know that, thanks a lot for that info.